src/rubrix/client/dataset.py

```diff
@@ -0,0 +1,350 @@
+#  coding=utf-8
+#  Copyright 2021-present, the Recognai S.L. team.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import logging
+from typing import Dict, List, Optional, Union
+
+import pandas as pd
+from pkg_resources import parse_version
+
+from rubrix.client.models import (
+    Record,
+    Text2TextRecord,
+    TextClassificationRecord,
+    TokenAttributions,
+    TokenClassificationRecord,
+)
+from rubrix.client.sdk.datasets.models import TaskType
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class Dataset:
+    """Dataset is a container class for Rubrix records.
+
+    Args:
+        records: A list of Rubrix records.
+
+    Raises:
+        MixedRecordTypesError: When the provided list contains more than one type of record.
+
+    Examples:
+        Passing in a list of records:
+        >>> import rubrix as rb
+        >>> records = [
+        ...     rb.TextClassificationRecord(inputs="example"),
+        ...     rb.TextClassificationRecord(inputs="another example"),
+        ... ]
+        >>> dataset = rb.Dataset(records)
+
+        Appending records to the dataset:
+        >>> import rubrix as rb
+        >>> dataset = Dataset()
+        >>> for text in ["example", "another example"]:
+        ...     datasets.append(rb.TextClassificationRecord(inputs=text))
+
+        Indexing into the dataset:
+        >>> dataset[0]
+        ... rb.TextClassificationRecord(inputs={"text": "example"})
+        >>> dataset[0] = rb.TextClassificationRecord(inputs="replaced example")
+
+        Looping over the dataset:
+        >>> assert len(dataset) == 2
+        >>> for record in dataset:
+        ...     print(record)
+    """
+
+    def __init__(self, records: Optional[List[Record]] = None):
```

Comment:
Yes, I thought about having an (abstract or generic) parent class, but I could not really figure out how to make it work:
1. I don't want the user to import different "types" of `Dataset`s when creating a new one, we already have the different record types for that. 
```python
# for example always
Dataset.from_datasets(...)
# instead of
TextClassificationDataset.from_datasets(...)
TokenClassificationDataset.from_datasets(...)
```
2. I would like to enable a similar "list appending" flow that we are using a lot right now in the tutorial:
```python
empty_dataset = Dataset()
for i in dataframe.itertuples():
    record = rb.TextClassificationRecord(...)
    empty_dataset.append(record)
```

Regarding pydantic, i like the idea of delegating the type checks!! Do you know if it is possible to have private attributes type checked? I don't want to expose the `_records` attribute openly, to prevent something like `Dataset.records.append(whateverthef***iwant)`.

Reply:
> I don't want the user to import different "types" of Datasets when creating a new one, we already have the different record types for that.

I changed my mind, the implementation will be much cleaner with different Dataset types, maybe we could use `Dataset.from_datasets` as factory method, but in principle the user should use the different Dataset types directly.

Reply:
Indeed, for private properties, the use of pydantic does not have much impact. You could use a generic class that uses some pydantic parsers for type validation. As shown in the code snippet

````python
class A(Generic[R]):
    def __init__(self, data: List[R]):
        self.__item_type__ = get_args(self.__class__.__orig_bases__[0])[0]
        self.__records__ = []
        for d in data:
            self.append_record(d)

    def append_record(self, data: R):
        self.__records__.append(parse_obj_as(self.__item_type__, data))

    @property
    def records(self) -> Iterable[R]:
        return tuple(self.__records__)

class B(A[int]):
    pass


r = B(data=[1, 23])
r.append_record("lolo") # This will raise a ValidationError

````